addons/binding/org.openhab.binding.opensprinkler/src/main/java/org/openhab/binding/opensprinkler/discovery/OpenSprinklerDiscoveryService.java

```diff
@@ -0,0 +1,151 @@
+import java.util.concurrent.TimeUnit;
+
+import org.apache.commons.net.util.SubnetUtils;
+import org.eclipse.smarthome.config.discovery.AbstractDiscoveryService;
+import org.eclipse.smarthome.config.discovery.DiscoveryResult;
+import org.eclipse.smarthome.config.discovery.DiscoveryResultBuilder;
+import org.eclipse.smarthome.core.thing.ThingTypeUID;
+import org.eclipse.smarthome.core.thing.ThingUID;
+import org.openhab.binding.opensprinkler.config.OpenSprinklerConfig;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * The {@link OpenSprinklerDiscoveryService} class allow manual discovery of
+ * OpenSprinkler devices for OpenHab.
+ *
+ * @author Chris Graham - Initial contribution
+ */
+public class OpenSprinklerDiscoveryService extends AbstractDiscoveryService {
+    private final Logger logger = LoggerFactory.getLogger(OpenSprinklerDiscoveryService.class);
+    private final static Set<ThingTypeUID> SUPPORTED_THING_TYPES_UIDS = new HashSet<ThingTypeUID>(
+            Arrays.asList(OPENSPRINKLER_THING, OPENSPRINKLERPI_THING));
+
+    private ExecutorService discoverySearchPool;
+
+    public OpenSprinklerDiscoveryService() {
+        super(SUPPORTED_THING_TYPES_UIDS, DISCOVERY_DEFAULT_TIMEOUT_RATE, DISCOVERY_DEFAULT_AUTO_DISCOVER);
+    }
+
+    @Override
+    public Set<ThingTypeUID> getSupportedThingTypes() {
+        return SUPPORTED_THING_TYPES_UIDS;
+    }
+
+    @Override
+    protected void startScan() {
+        logger.debug("Starting discovery of OpenSprinkler devices.");
+
+        try {
+            List<String> ipList = getIpAddressScanList();
+
+            discoverySearchPool = Executors.newFixedThreadPool(DISCOVERY_THREAD_POOL_SIZE);
+
+            for (String ip : ipList) {
+                discoverySearchPool.execute(new OpenSprinklerDiscoveryThread(this, ip));
+            }
+
+            discoverySearchPool.shutdown();
+
+            try {
+                discoverySearchPool.awaitTermination(DISCOVERY_THREAD_POOL_SHUTDOWN_WAIT_TIME_SECONDS,
+                        TimeUnit.SECONDS);
+            } catch (InterruptedException e) {
+                logger.error("OpenSprinkler discover service reached the timeout value waiting to shutdown.");
+            }
+
+            discoverySearchPool = null;
+        } catch (Exception exp) {
+            logger.debug("OpenSprinkler discovery service encountered an error while scanning for devices: "
+                    + exp.getMessage());
+        }
+
+        logger.debug("Completed discovery of OpenSprinkler devices.");
+    }
+
+    /**
+     * Create a new Thing in OpenHab with an IP address given. Uses default port and password.
+     *
+     * @param ip IP address of the OpenSprinkler device as a string.
+     */
+    public void submitDiscoveryResults(String ip) {
+        ThingUID uid = new ThingUID(OPENSPRINKLER_THING, ip.replace('.', '_'));
+
+        if (uid != null) {
+            OpenSprinklerConfig properties = new OpenSprinklerConfig();
```

Comment:
No need to create the Config here. Either create a Map right here or use withProperty() on the builder for all 4 values.